src/cpp/session/modules/rmarkdown/SessionRmdNotebook.cpp

```diff
@@ -0,0 +1,670 @@
+   if (consoleId == s_consoleChunkId)
+   {
+      if (s_consoleConnected) 
+         return;
+      connectConsole();
+   }
+   else if (s_consoleConnected)
+   {
+      // some other console is connected; disconnect ours
+      disconnectConsole();
+   }
+}
+
+Error handleChunkOutputRequest(const http::Request& request,
+                               http::Response* pResponse)
+{
+   // uri format is: /chunk_output/<doc-id>/...
+   
+   // split URI into pieces, extract the document ID, and remove that part of
+   // the URI
+   std::vector<std::string> parts = algorithm::split(request.uri(), "/");
+   if (parts.size() < 4) 
+      return Success();
+   std::string docId = parts[2];
+   for (int i = 0; i < 3; i++)
+      parts.erase(parts.begin());
+
+   std::string path;
+   Error error = source_database::getPath(docId, &path);
+   if (error)
+      return error;
+   FilePath target = chunkCacheFolder(path, docId).complete(
+         algorithm::join(parts, "/"));
+
+   if (parts[0] == kChunkLibDir)
+   {
+      // if a reference to the chunk library folder, we can reuse the contents
+      pResponse->setCacheableFile(target, request);
+   }
+   else
+   {
+      // otherwise, we don't want the output to be cached (so we get fresh
+      // chunk output every time)
+      pResponse->setNoCacheHeaders();
+      pResponse->setFile(target, request);
+   }
+
+   return Success();
+}
+
+// called by the client to set the active chunk console
+Error setChunkConsole(const json::JsonRpcRequest& request,
+                      json::JsonRpcResponse*)
+{
+
+   std::string docId, chunkId;
+   Error error = json::readParams(request.params, &docId, &chunkId);
+   if (error)
+      return error;
+
+   s_consoleChunkId = chunkId;
+   s_consoleDocId = docId;
+   if (s_activeConsole == chunkId)
+      connectConsole();
+
+   return Success();
+}
+
+// called by client to execute a single chunk in a document
+Error executeInlineChunk(const json::JsonRpcRequest& request,
+                         json::JsonRpcResponse*)
+{
+   std::string docPath, docId, chunkId, chunkOptions, content;
+   Error error = json::readParams(request.params, &docPath, &docId, &chunkId, 
+         &chunkOptions, &content);
+   if (error)
+      return error;
+
+   // ensure we have a place to put the output
+   FilePath chunkOutput = chunkOutputPath(docPath, docId, chunkId);
+   error = chunkOutput.parent().ensureDirectory();
+   if (error)
+      return error;
+
+   // ensure we have a library path
+   FilePath chunkLibDir = chunkCacheFolder(docPath, docId).complete(
+         kChunkLibDir);
+   error = chunkLibDir.ensureDirectory();
+   if (error)
+      return error;
+
+   FilePath headerHtml = options().rResourcesPath().complete("notebook").
+      complete("in_header.html");
+
+   // render the contents to the cached folder
+   std::string errorMsg;
+   error = r::exec::RFunction(".rs.executeSingleChunk", chunkOptions, content,
+         chunkLibDir.absolutePath(),
```

Comment:
No, we only need that when we're forming a string which will be evaluated in the shell (e.g. a command line arg to an async R process). But we *were* missing encoding markers on the path strings on the R side of this call, so I've added those.